### PR TITLE
Adding OAUTHLIB_INSECURE_TRANSPORT and LOGIN_DISABLED values

### PR DESCRIPTION
**Pull request by jainsushma**
Please make sure you are raising this PR against your own repository and not the original. If it says https://github.com/CorndelWithSoftwire in the search bar right now, you are in the wrong place!

Please also don't forget to submit the PR's url to Aptem afterwards (by clicking on the corresponding exercise component, pasting the url in the textbox that appears and then pressing the finish button).

### CODE REVIEW

File: todo_app/app.py [hunk 1]
```diff
@@ -3,26 +3,99 @@
...
+        tokenUrl, headers, body = githubClient.prepare_token_request("https://github.com/login/oauth/access_token", code=callbackCode)
+        headers["Accept"] = "application/json"
+        r = requests.post(
+            tokenUrl,
+            headers=headers,
+            data=body,
+            auth=(clientId, clientSecret),
+        )
+        params=githubClient.parse_request_body_response(r.text)
+        print(params['access_token'])
```
Comment (JackMeadDev) [inline todo_app/app.py]: This is a sensitive token, so while it's very helpful for debugging we should remove it from our logs afterwards!

File: todo_app/app.py [hunk 2]
```diff
@@ -3,26 +3,99 @@
...
+
+    def validate_user_role(func):
+        @wraps(func)
+        def wrapTheFunction(*args, **kwargs):
+            if loginDisabled or (current_user.role == "reader"):
+                if loginDisabled:
+                    return func(*args, **kwargs)
+                elif (current_user.role == "reader"):    
+                    return render_template("error.html", error="Insufficient User Rights") 
+                return func(*args, **kwargs)
```
Comment (JackMeadDev) [inline todo_app/app.py]: I think you've got a mismatch here - if security is enabled and the role is "writer" then the user doesn't trigger the outer if, and nothing happens - so the wrapped function doesn't execute at all (which leads to errors)

Removing the outer if statement doesn't change most cases, but does fix the error:
```suggestion
            if loginDisabled:
                return func(*args, **kwargs)
            elif (current_user.role == "reader"):    
                return render_template("error.html", error="Insufficient User Rights") 
            return func(*args, **kwargs)    
```

Having said that, note that loginDisabled treats the user as a writer, so we could simplify further to:
```suggestion
            if (current_user.role == "reader"):    
                return render_template("error.html", error="Insufficient User Rights") 
            return func(*args, **kwargs)    
```

File: README.md [hunk 3]
```diff
@@ -128,4 +128,7 @@ CLIENT=[Your Connection String Here]
 
 2. Add CLIENT secret to Github
 3. Add CLIENT and some new DB eg 'PRODDB' as DBNAME as Heroku config variables
-```
\ No newline at end of file
+```
+
+## Linking to OAuth
+To-Do App is using GitHub for linking to default OAuth provider to manage user access. Users will now be redirected to github to sign and access the application. You will need to register your application with your chosen Auth provider and populate your .env file with the enviroment variables found in the .env.template with the values taken from Github OAuth. Users with "read" access will not be allowed to add, update or delete the tasks in the app.
```
Comment (JackMeadDev) [inline README.md]: Minor, but as you say we're using GitHub, and choosing an alternative Auth Provider won't _just work_ - we'd have to change the code to allow for that!

File: todo_app/app.py [hunk 4]
```diff
@@ -3,26 +3,99 @@
...
+        githubRedirect = githubClient.prepare_request_uri("https://github.com/login/oauth/authorize")
+        print(githubRedirect)
+        return redirect(githubRedirect)
+
+    @login_manager.user_loader 
+    def load_user(user_id):
+        user = User(user_id)
+        return user
+
+    def validate_user_role(func):
```
Comment (JackMeadDev) [inline todo_app/app.py:63]: Great use of a decorator here 😄

**Review by JackMeadDev — APPROVED** [review]
Great work - I was forced to authenticate through Github and wasn't able to get past your security as a read-only user, and this is a clean solution to the problem! There was an error however when I tried to change things as a writer - I suspect you changed something late on that broke that use case when accounting for `loginDisabled` - see my comment below for how to fix that. Also worth noting is that your deployed Heroku app doesn't have appropriate environment variables set up so the auth journey fails currently - you'll need to set up a new OAuth app which is configured for that web address & set  the environment variables to get that working

Once you have those ready, this looks good to be merged.

File: todo_app/app.py (hunk 1 above)
Comment (jainsushma) [reply todo_app/app.py]: Done!

File: todo_app/app.py (hunk 2 above)
Comment (jainsushma) [reply todo_app/app.py]: Done

File: README.md (hunk 3 above)
Comment (jainsushma) [reply README.md]: Done

File: todo_app/app.py (hunk 4 above)
Comment (jainsushma) [reply todo_app/app.py:63]: Thanks

File: todo_app/app.py (hunk 2 above)
Comment (jainsushma) [reply todo_app/app.py]: It seems need to make use of loginDisabled to keep e2e tests passing in Github actions build. Hope following changes are ok
`  if (not loginDisabled and current_user.role == "reader"):    
                return render_template("error.html", error="Insufficient User Rights") 
            return func(*args, **kwargs) `

**Comment by jainsushma** [comment]
@JackMead , I have added 
`OAUTHLIB_INSECURE_TRANSPORT=1
LOGIN_DISABLED=False`

in config variables for Heroku. Still https://heroku-app-devops.herokuapp.com/ is broken. How do you suggest to make it working?

File: todo_app/app.py (hunk 2 above)
Comment (JackMeadDev) [reply todo_app/app.py]: Ah that makes sense - `current_user` isn't set if login is disabled. That sounds reasonable to me 👍

**Comment by JackMeadDev** [comment]
If you try the link, you'll see it redirects you to:
https://github.com/login/oauth/authorize?response_type=code&client_id=None

Note the "client_id=None" specifically is the hint we need - there are two more environment variables that you added as part of this ticket:
```
CLIENTID=[Your Client ID Here]
CLIENTSECRET=[Your Client Secret Here]
```
It's those two that you'll need to set, but having set up a new OAuth app in Github that uses your Heroku URL rather than localhost 🙂

**Pull request merged**